beacon_node/beacon_chain/src/graffiti_calculator.rs

```diff
@@ -0,0 +1,399 @@
+                let Some(execution_layer) = self.execution_layer.as_ref() else {
+                    // Return default graffiti if there is no execution layer. This
+                    // shouldn't occur if we're actually producing blocks.
+                    crit!(self.log, "No execution layer available for graffiti calculation during block production!");
+                    return default_graffiti;
+                };
+
+                // The engine version cache refresh service ensures this will almost always retrieve this data from the
+                // cache instead of making a request to the execution engine. A cache miss would only occur if lighthouse
+                // has recently started or the EL recently went offline.
+                let engine_versions = match execution_layer
+                    .get_engine_version(Some(
+                        self.epoch_duration * ENGINE_VERSION_AGE_LIMIT_EPOCH_MULTIPLE,
+                    ))
+                    .await
+                {
+                    Ok(engine_versions) => engine_versions,
+                    Err(el_error) => {
+                        warn!(self.log, "Failed to determine execution engine version for graffiti"; "error" => ?el_error);
+                        return default_graffiti;
+                    }
+                };
+
+                let Some(engine_version) = engine_versions.first() else {
+                    // Got an empty array which indicates the EL doesn't support the method
+                    debug!(
+                        self.log,
+                        "Using default lighthouse graffiti: EL does not support {} method",
+                        ENGINE_GET_CLIENT_VERSION_V1;
+                    );
+                    return default_graffiti;
+                };
+                if engine_versions.len() != 1 {
+                    // More than one version implies lighthouse is connected to
+                    // an EL multiplexer. We don't support modifying the graffiti
+                    // with these configurations.
+                    warn!(
+                        self.log,
+                        "Execution Engine multiplexer detected, using default graffiti"
+                    );
+                    return default_graffiti;
+                }
+
+                let lighthouse_commit_prefix = CommitPrefix::try_from(lighthouse_version::COMMIT_PREFIX.to_string())
+                    .unwrap_or_else(|error_message| {
+                        // This really shouldn't happen but we want to definitly log if it does
+                        crit!(self.log, "Failed to parse lighthouse commit prefix"; "error" => error_message);
+                        CommitPrefix("00000000".to_string())
+                    });
+
+                engine_version.calculate_graffiti(lighthouse_commit_prefix)
+            }
+        }
+    }
+}
+
+pub fn start_engine_version_cache_refresh_service<T: BeaconChainTypes>(
```

Comment:
I'm not sure why we need to maintain a service to keep updating the version.
Why not instead do it once on startup, then do it in `BeaconChain::prepare_beacon_proposer`. Updating something that doesn't change repeatedly seems quite wasteful to me

Reply:
We could do it there, but then it would happen right in window between notifying the EL that there's a proposal & actually proposing the block, which is a critical time. The whole reason I wrote it this way to avoid extra calls to the EL within that critical time.

Presumably we could have a service that only caches this result when we detect that we are connected to a validator which will have a proposal within the next epoch, but that code would surely be more complex that this service.

Reply:
I don't think that making a single call to an EL that returns a pre set value should be very intensive, but your point is fair. Better to not cram more calls than necessary just before a proposal.